Clean up room connection status

nandito opened the pull request:
### Summary

Clean up SDK's api surface before the beta release.

### Changes

* Remove `accepted` state from `roomConnectionStatus`
      > This state is not used for anything, it should be safe to remove.
* Rename `rejected` state to `knock_rejected`
      > Use a more descriptive status name when a knock request gets rejected on a locked room.
* Rename `roomConnectionStatus` -> `connectionStatus`
* Remove `mostRecentChatMessage`
* Remove `startScreenshareError`
* Replace `isStartingScreenshare` with `localScreenshareStatus` and fix the missing state reset when screen share stops
* Make roomConnectionOptions optional
       > Use default localMediaConstraints' audio, video: true.
* Use the default media constraints in the related story
       > `useLocalMedia` already has a default option and the same value was
passed in the story. This commit removes this unnecessary param.
* Do not expose `newJoiner` and `streams` on `remoteParticipant`
       > These are internal props, no need to expose them to in the useRoomConnection's `state`.

### Test plan

knock:
1. Use the sdk with a locked room
2. Join the room with a host key
3. Knock the room with a guest participant
4. Reject the knock with the host => room connection status should be `knock_rejected`

screenshare status:
1. start/stop screenshares and inspect the changes of the state's `localScreenshareStatus`
2. initially it is `undefined`
3. starting a screenshare => `starting`
4. when screenshare is in progress => `active`
5. stop screenshare either by the `stopScreenshare` action or by the browser's native alert => `undefined` again

useRoomConnection's default options:
1. use the sdk's `useRoomConnection` with a roomUrl only, like `useRoomConnection(roomUrl)`
2. it should connect with the default media constraints

`newJoiner` and `streams` props in `remoteParticipant`
1. `console.log` the `state` or `state.remoteParticipants` in your demo app
2. join the room with 2 participants
3. check the logs: `newJoiner` and `streams` props shouldn't be there.

Make sure your demo app works after migrating to this version 😇 .

nandito committed:
Rename `rejected` status to `knock_rejected`

nandito committed:
Remove `accepted` state from room connection status

This state is not used for anything, it should be safe to remove.

nandito committed:
Rename `roomConnectionStatus` to `connectionStatus`

nandito committed:
Remove `mostRecentChatMessage`

nandito committed:
Remove `startScreenshareError`

nandito committed:
Replace `isStartingScreenshare` with `localScreenshareStatus`

... and fix the missing state reset when screen share stops.

nandito committed:
Make `roomConnectionOptions` optional

Use default localMediaConstraints' audio, video: true.

nandito committed:
Use the default media constraints in the related story

`useLocalMedia` already has a default option and the same value was
passed in the story. This commit removes this unnecessary param.

nandito committed:
Omit `newJoiner`,`streams`, `addStream`, `removeStream` from `remoteParticipants`

These props are only used internally, so no need to add them to the
useRoomConnection state.

nandito committed:
Bump version tag to beta1

kevinwhereby commented on src/lib/react/useRoomConnection.ts:
```diff
@@ -11,10 +11,12 @@ import RoomConnection, {
 } from "../RoomConnection";
 import { LocalParticipantState, RemoteParticipantState, Screenshare, WaitingParticipant } from "../RoomParticipant";
 
+type LocalScreenshareStatus = undefined | "starting" | "active";
```
how do we feel about moving the optional-ness to the object property like
```typescript
type LocalScreenshareStatus = "starting" | "active"
export interface RoomConnectionState {
    ...rest;
    localScreenshareStatus?: LocalScreenshareStatus;
}
```

havardholvik replied:
I dont have strong opinions here, but I see we are doing the same thing with recording and streaming state. If there is no state, the property is undefined

kevinwhereby commented on src/lib/react/useRoomConnection.ts:
```diff
@@ -379,7 +379,17 @@ export type RoomConnectionRef = {
+
+export function useRoomConnection(
+    roomUrl: string,
+    roomConnectionOptions = defaultRoomConnectionOptions
```
I wonder if we should move the default options into the body so people could include some options without localMediaConstraints?

something like 
```typescript
export function useRoomConnection(
    roomUrl: string,
    roomConnectionOptions: UseRoomConnectionOptions;
) {
    const [roomConnection] = useState<RoomConnection>(
        () =>
            new RoomConnection(roomUrl, {
                ...defaultRoomConnectionOptions,
                ...roomConnectionOptions,
                localMedia: roomConnectionOptions?.localMedia?._ref || undefined,
            })
      );
}

Though I can see how that could be confusing for the end user

nandito replied:
Yeah, both solutions have pros and cons. 🤔 Personally, I find the currently committed one to be more straightforward to use. I mean, if I provide options, only those options will be used. However, if you feel that your suggestions fit better, I'm happy to change it. I can see its benefits too. 😅

kevinwhereby replied:
nah lets not bother, thinking about it more I agree with you

kevinwhereby commented on src/lib/RoomConnection.ts:
```diff
@@ -426,7 +445,11 @@ export default class RoomConnection extends TypedEventTarget {
         this.remoteParticipants = [...this.remoteParticipants, remoteParticipant];
         this._handleAcceptStreams([remoteParticipant]);
-        this.dispatchEvent(new CustomEvent("participant_joined", { detail: { remoteParticipant } }));
+        this.dispatchEvent(
```
Should we move the state conversions up to the listeners in the `useRoomConnection` hook? If we ever want to reuse this `RoomConnection` class elsewhere we might want access to the full class, and it feels like a task that should be handled where we handle the rest of the react state

nandito replied:
I'm not sure if I understand correctly your point here. I only omit the props in the CustomEvent's payload, nothing changes in the `RoomConnection` class. 🤔 

Omitting the props here satisfies the TS better in this case, as it expects `RemoteParticipantState` in the payload, not `RemoteParticipant`.

```ts
var CustomEvent: new <{
    remoteParticipant: RemoteParticipantState;
}>(type: string, eventInitDict?: CustomEventInit<{
    remoteParticipant: RemoteParticipantState;
}> | undefined) => CustomEvent<...>
```

kevinwhereby replied:
So looking at the `RoomJoinedEvent` defined in `RoomConnection`, we see it's payload uses `RemoteParticipant`s
```typescript
// RoomConnection.ts line 71
export type RoomJoinedEvent = {
    localParticipant: LocalParticipant;
    remoteParticipants: RemoteParticipant[];
    waitingParticipants: WaitingParticipant[];
};
// line 138
export interface RoomEventsMap {
    ...rest;
    room_joined: (e: CustomEvent<RoomJoinedEvent>) => void;
}
// line 345
this.signalSocket.on("room_joined", this._handleRoomJoined.bind(this));
```

but then when we dispatch this `room_joined` event, we're converting to `RemoteParticipantState` prematurely, like so
```typescript
// RoomConnection.ts line 531
private _handleRoomJoined(event: SignalRoomJoinedEvent) {
    ...
    this.dispatchEvent(
        new CustomEvent("room_joined", {
            detail: {
                localParticipant: this.localParticipant,
                    remoteParticipants: this.remoteParticipants.map(
                        convertRemoteParticipantToRemoteParticipantState
                    ),
                    waitingParticipants: knockers.map((knocker) => {
                        return { id: knocker.clientId, displayName: knocker.displayName } as WaitingParticipant;
                    }),
                },
            })
        );
}
```
I think we need to move the state conversion up inside `useRoomConnection` like so:
```typescript
// useRoomConnection.ts line 475
createEventListener("room_joined", (e) => {
    const { localParticipant, remoteParticipants, waitingParticipants } = e.detail;
    dispatch({
        type: "ROOM_JOINED",
        payload: { 
            localParticipant, 
            remoteParticipants: remoteParticipants.map(convertRemoteParticipantToRemoteParticipantState), 
            waitingParticipants,
        },
    });
}),
```

hope that makes sense 😄

kevinwhereby replied:
you can see the error in typescript if you give the created custom event the `RoomJoinedEvent` type:
```typescript
            const customEvent: RoomJoinedEvent = new CustomEvent("room_joined", {
                detail: {
                    localParticipant: this.localParticipant,
                    remoteParticipants: this.remoteParticipants.map(convertRemoteParticipantToRemoteParticipantState),
                    waitingParticipants: knockers.map((knocker) => {
                        return { id: knocker.clientId, displayName: knocker.displayName } as WaitingParticipant;
                    }),
                },
            });
            this.dispatchEvent(customEvent);
```

...which probably means we need to type these events before dis to avoid this kind of error in the first place 😄

nandito replied:
Yes, this makes sense. Thanks a lot for the well detailed explanation! 🙏

kevinwhereby commented on src/lib/RoomParticipant.ts:
```diff
@@ -79,7 +79,10 @@ export class RemoteParticipant extends RoomParticipant {
 }
 
-export type RemoteParticipantState = Omit<RemoteParticipant, "updateStreamState">;
+export type RemoteParticipantState = Omit<
```
though with the definition of the state type here, maybe it is something the RoomConnection handles... I'm not fussed either way really just thought I'd share my thoughts 😄

havardholvik replied:
Agree this should be exported from the useRoomConnection module

nandito committed:
Move RemoteParticipantState typedef to useRoomConnection

nandito committed:
Clean remoteParticipants state in the hook

kevinwhereby reviewed (approved):
awsome!

kevinwhereby commented:
Currently elbow deep in typescript trying to make a constructor for these events that types them automatically 😂 😭 
```typescript
type RoomEventKey = keyof RoomEventsMap;
type ArgType<T> = T extends (arg: infer U) => unknown ? U : never;
type RoomEventHandler<T extends RoomEventKey> = RoomEventsMap[T];
type RoomEvent<T extends RoomEventKey> = ArgType<RoomEventHandler<T>>;
type RoomEventPayload<T extends RoomEventKey> = RoomEvent<T> extends CustomEvent<infer U> ? U : never;

function constructCustomEvent<T extends RoomEventKey>(
    eventType: T,
    eventInitDict: { detail: RoomEventPayload<T> }
): RoomEvent<T> {
    return new CustomEvent(eventType, eventInitDict);
}
```

kevinwhereby committed:
create RoomEvent/LocalMediaEvent class to type event construction

thin wrappers around CustomEvent that will enforce our payload types on
the created events

kevinwhereby committed:
fix type errors arising from typed event payloads

havardholvik committed:
Bump jslib-media to avoid .getStats() error

kevinwhereby committed:
rename RoomEvent to RoomConnectionEvent

kevinwhereby committed:
Merge pull request #103 from whereby/kevinhanna/room-event-typing

Type custom event payloads

kevinwhereby committed:
separate module imports

havardholvik committed:
Remove exports of low-level classes

We dont want to expose the LocalMeda and RoomConnection classes just
yet, so leave them out of the exports.

kevinwhereby committed:
Merge pull request #105 from whereby/kevinhanna/separate-module-exports

separate module imports

havardholvik committed:
Fixed type exports
